africanus/model/spi/component_spi.py

```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from collections import namedtuple
+from ...util.docs import DocstringTemplate
+
+import numba
+import numpy as np
+
+
+@numba.jit(nopython=True, nogil=True, cache=True)
+def _fit_spi_components_impl(data, weights, freqs, freq0, out,
+                             jac, ncomps, nfreqs, tol, maxiter):
+    w = freqs/freq0
+    for comp in range(ncomps):
+        eps = 1.0
+        k = 0
+        alphak = out[0, comp]
+        i0k = out[2, comp]
+        while eps > tol and k < maxiter:
+            alphap = alphak
+            i0p = i0k
+            jac[1, :] = w**alphak
+            model = i0k*jac[1, :]
+            jac[0, :] = model * np.log(w)
+            residual = data[comp] - model
+            lik = 0.0
+            hess00 = 0.0
+            hess01 = 0.0
+            hess11 = 0.0
+            jr0 = 0.0
+            jr1 = 0.0
+            for v in range(nfreqs):
+                lik += residual[v] * weights[v] * residual[v]
+                jr0 += jac[0, v] * weights[v] * residual[v]
+                jr1 += jac[1, v] * weights[v] * residual[v]
+                hess00 += jac[0, v] * weights[v] * jac[0, v]
+                hess01 += jac[0, v] * weights[v] * jac[1, v]
+                hess11 += jac[1, v] * weights[v] * jac[1, v]
+            det = hess00 * hess11 - hess01**2
+            alphak = alphap + (hess11 * jr0 - hess01 * jr1)/det
+            i0k = i0p + (-hess01 * jr0 + hess00 * jr1)/det
+            eps = np.maximum(np.abs(alphak - alphap), np.abs(i0k - i0p))
+            k += 1
+        if k >= maxiter:
+            print("Warning - max iterations exceeded for component ", comp)
```

Comment:
Does this ever occur, given the conditional in the while loop? If not, it can probably be removed.